### PR TITLE
Move additional dependencies from private getters to constructor - Magento_Catalog

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
### Description (*)

This PR moves additionally introduced dependencies from private getter methods into constructor in Magento_Catalog module.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php [hunk 1]
```diff
@@ -141,6 +143,7 @@ public function __construct(
         ProductLinks $productLinks,
         Js $jsHelper,
         Date $dateFilter,
+        LinkResolver $linkResolver,
         CustomOptionFactory $customOptionFactory = null,
```
Comment (orlangur) [inline app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php]: Please add all new dependencies in a backward-compatible manner - similar to other added in this class:
1) new dependency is last
2) `null` default value
3) initialized in constructor when it was not passed

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php (hunk 1 above)
Comment (Bartlomiejsz) [reply app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php]: I was sure that it should be done this way but in few of my previous PRs it was mentioned that for 2.4 release proper way is without ObjectManager, and I had to change it for such like here. So which is correct version?

Comment (Bartlomiejsz) [reply app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php]: @orlangur check https://github.com/magento/magento2/pull/26269 as an example

Comment (Bartlomiejsz) [reply app/code/Magento/Catalog/Controller/Adminhtml/Product/Initialization/Helper.php]: Hi @orlangur, also check this one please https://github.com/magento/magento2/commit/4e6f6ae42cc6109a02d35d0363b1cad0bbac66e1 this was commited by Magento core and this is pretty much the same I did here

**Comment by Bartlomiejsz** [comment]
Closing this one to open new PR as draft, since Catalog requires serious amount of modifications

**Pull request closed without merging**